// tdesign-react/es/date-picker/panel/DatePickerPanel.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-08260abc.js';
import React from 'react';
import classNames from 'classnames';
import dayjs from 'dayjs';
import useConfig from '../../_util/useConfig.js';
import PanelContent from './PanelContent.js';
import ExtraContent from './ExtraContent.js';
import useTableData from './useTableData.js';
import useDisableDate from '../hooks/useDisableDate.js';
import '../../_chunks/dep-e3ed7551.js';
import '../../_chunks/dep-12690e1d.js';
import '../../_chunks/dep-ac0cc4dd.js';
import '../../_chunks/dep-e3ab6b0d.js';
import '../../_chunks/dep-cc4f2bae.js';
import '../../_chunks/dep-531703d1.js';
import '../../_chunks/dep-21baa91f.js';
import '../../_chunks/dep-4705ff31.js';
import '../../_chunks/dep-38424935.js';
import '../../_chunks/dep-ad150068.js';
import '../../_chunks/dep-d5f141ac.js';
import '../../_chunks/dep-1b27e5c2.js';
import '../../_chunks/dep-6963df41.js';
import '../../_chunks/dep-8fb1642c.js';
import '../../_chunks/dep-bcb59ae2.js';
import '../../_chunks/dep-12718f01.js';
import '../../_chunks/dep-09c318fb.js';
import '../../_chunks/dep-048c506f.js';
import '../../_chunks/dep-febf8cf9.js';
import '../../_chunks/dep-074143e4.js';
import '../../_chunks/dep-a1fd625f.js';
import '../../_chunks/dep-f5ab8a5c.js';
import '../../_chunks/dep-67009882.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../config-provider/zh_CN_config.js';
import '../base/Header.js';
import '../../_chunks/dep-bd2639ad.js';
import '../../_chunks/dep-0e5be0b2.js';
import '../../_chunks/dep-744640f5.js';
import '../../locale/LocalReceiver.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../select/index.js';
import '../../select/base/Select.js';
import '../../_chunks/dep-8ff4a741.js';
import '../../_chunks/dep-71b3c60f.js';
import '../../_chunks/dep-084711a9.js';
import '../../_chunks/dep-863396c0.js';
import '../../_chunks/dep-e8f6f682.js';
import '../../tag/index.js';
import '../../tag/Tag.js';
import 'tdesign-icons-react';
import '../../_util/noop.js';
import '../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../tag/CheckTag.js';
import '../../hooks/useControlled.js';
import '../../_chunks/dep-0db6b692.js';
import '../../_chunks/dep-6d2697ef.js';
import '../../_chunks/dep-42701b6c.js';
import '../../tag/defaultProps.js';
import './style/css.js';
import '../../tag/type.js';
import '../../select/util/helper.js';
import '../../select/base/OptionGroup.js';
import '../../select/defaultProps.js';
import '../../select/base/Option.js';
import '../../_chunks/dep-1feb6aff.js';
import '../../_chunks/dep-60c49e74.js';
import '../../_util/useRipple.js';
import '../../_util/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';
import '../../common/FakeArrow.js';
import '../../loading/index.js';
import '../../loading/Loading.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../loading/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';
import '../../loading/type.js';
import '../../select-input/index.js';
import '../../select-input/SelectInput.js';
import '../../popup/index.js';
import '../../popup/Popup.js';
import 'react-transition-group';
import 'react-popper';
import '../../_util/composeRefs.js';
import '../../popup/hooks/useTriggerProps.js';
import '../../_util/useClickOutside.js';
import '../../popup/utils/getTransitionParams.js';
import '../../_util/useMutationObserver.js';
import '../../_chunks/dep-0a541e21.js';
import '../../_chunks/dep-5ed520e8.js';
import '../../_util/useWindowSize.js';
import '../../popup/defaultProps.js';
import '../../popup/type.js';
import '../../select-input/useSingle.js';
import '../../_chunks/dep-14dc871a.js';
import '../../_chunks/dep-d8cdbac1.js';
import '../../_chunks/dep-b9ff0943.js';
import '../../_chunks/dep-3cff266a.js';
import '../../_chunks/dep-43a6fda3.js';
import '../../input/index.js';
import '../../input/Input.js';
import '../../_util/helper.js';
import '../../_chunks/dep-9be29560.js';
import '../../input/InputGroup.js';
import '../../input/defaultProps.js';
import '../../input/type.js';
import '../../select-input/useMultiple.js';
import '../../tag-input/index.js';
import '../../tag-input/TagInput.js';
import '../../_util/useDragSorter.js';
import '../../tag-input/useTagScroll.js';
import '../../tag-input/useTagList.js';
import 'lodash';
import '../../tag-input/useHover.js';
import '../../tag-input/defaultProps.js';
import '../../tag-input/type.js';
import '../../select-input/useOverlayStyle.js';
import '../../select-input/defaultProps.js';
import '../../select-input/type.js';
import '../../select/base/PopupContent.js';
import '../../select/type.js';
import '../../common/Jumper.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../button/defaultProps.js';
import '../../button/type.js';
import '../base/Table.js';
import '../base/Cell.js';
import '../../_common/js/date-picker/utils-new.js';
import '../../_chunks/dep-3d09d1f3.js';
import '../../_chunks/dep-87214e71.js';
import '../../time-picker/panel/TimePickerPanel.js';
import '../../_chunks/dep-0b46b8b5.js';
import '../../_chunks/dep-c739cdde.js';
import '../../time-picker/hooks/useTimePickerTextConfig.js';
import '../../_common/js/time-picker/const.js';
import '../../_common/js/time-picker/utils.js';
import '../hooks/useFormat.js';
import '../base/Footer.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var DatePickerPanel = function DatePickerPanel(props) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix,
      globalDatePickerConfig = _useConfig.datePicker;

  var panelName = "".concat(classPrefix, "-date-picker__panel");
  var value = props.value,
      mode = props.mode,
      _props$format = props.format,
      format = _props$format === void 0 ? "YYYY-MM-DD" : _props$format,
      presetsPlacement = props.presetsPlacement,
      disableDateFromProps = props.disableDate,
      _props$firstDayOfWeek = props.firstDayOfWeek,
      firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
      style = props.style,
      className = props.className,
      year = props.year,
      month = props.month,
      _onClick = props.onClick;
  var disableDateOptions = useDisableDate({
    disableDate: disableDateFromProps,
    mode: mode,
    format: format
  });
  var tableData = useTableData(_objectSpread({
    year: year,
    month: month,
    mode: mode,
    start: value ? dayjs(value, format).toDate() : void 0,
    firstDayOfWeek: firstDayOfWeek
  }, disableDateOptions));
  var panelContentProps = {
    mode: mode,
    year: year,
    month: month,
    format: format,
    firstDayOfWeek: firstDayOfWeek,
    tableData: tableData,
    enableTimePicker: props.enableTimePicker,
    timePickerProps: props.timePickerProps,
    timeValue: props.timeValue,
    onMonthChange: props.onMonthChange,
    onYearChange: props.onYearChange,
    onJumperClick: props.onJumperClick,
    onCellClick: props.onCellClick,
    onCellMouseEnter: props.onCellMouseEnter,
    onCellMouseLeave: props.onCellMouseLeave,
    onTimePickerChange: props.onTimePickerChange
  };
  return /* @__PURE__ */React.createElement("div", {
    style: style,
    className: classNames(panelName, className, _defineProperty({}, "".concat(panelName, "--direction-row"), ["left", "right"].includes(presetsPlacement))),
    onClick: function onClick(e) {
      return _onClick === null || _onClick === void 0 ? void 0 : _onClick({
        e: e
      });
    }
  }, ["top", "left"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, _objectSpread(_objectSpread({}, props), {}, {
    selectedValue: value
  })) : null, /* @__PURE__ */React.createElement(PanelContent, _objectSpread({
    tableData: tableData
  }, panelContentProps)), ["bottom", "right"].includes(presetsPlacement) ? /* @__PURE__ */React.createElement(ExtraContent, _objectSpread(_objectSpread({}, props), {}, {
    selectedValue: value
  })) : null);
};

DatePickerPanel.displayName = "DatePickerPanel";
DatePickerPanel.defaultProps = {
  mode: "date",
  enableTimePicker: false,
  presetsPlacement: "bottom"
};

export { DatePickerPanel as default };
