// tdesign-react/lib/upload/themes/single-file.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var loading_index = require('../../loading/index.js');
var _util_useConfig = require('../../_util/useConfig.js');
require('../../loading/Loading.js');
require('../../_chunks/dep-a1655de1.js');
require('../../_chunks/dep-3cf9becb.js');
require('../../_chunks/dep-19841e1d.js');
require('../../_util/dom.js');
require('raf');
require('../../_chunks/dep-1d819290.js');
require('../../_chunks/dep-fcf32c81.js');
require('../../_chunks/dep-3405bda4.js');
require('../../_chunks/dep-c90cc28f.js');
require('../../_chunks/dep-b1a67107.js');
require('../../_util/easing.js');
require('../../common/Portal.js');
require('react-dom');
require('../../_chunks/dep-38a46f48.js');
require('../../_chunks/dep-8a149e9f.js');
require('../../_chunks/dep-3b69722d.js');
require('../../_chunks/dep-fde4d28a.js');
require('../../_chunks/dep-e266830c.js');
require('../../_chunks/dep-15a4424e.js');
require('../../_chunks/dep-c40d75ac.js');
require('../../_chunks/dep-910dcfa3.js');
require('../../_chunks/dep-633a5671.js');
require('../../_chunks/dep-9c339ffc.js');
require('../../_chunks/dep-ab43c961.js');
require('../../_chunks/dep-9cacaf01.js');
require('../../_chunks/dep-c72c7229.js');
require('../../_chunks/dep-8e0a771d.js');
require('../../_chunks/dep-542cc308.js');
require('../../_chunks/dep-c6637e11.js');
require('../../_common/js/global-config/locale/zh_CN.js');
require('../../config-provider/zh_CN_config.js');
require('../../loading/gradient.js');
require('../../_common/js/loading/circle-adapter.js');
require('../../_common/js/utils/set-style.js');
require('../../_common/js/utils/helper.js');
require('../../_chunks/dep-23535a52.js');
require('../../loading/defaultProps.js');
require('../../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var SingleFile = function SingleFile(props) {
  var _props$display = props.display,
      display = _props$display === void 0 ? "file" : _props$display,
      onRemove = props.onRemove,
      file = props.file,
      showUploadProgress = props.showUploadProgress;

  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var fileClass = classNames__default["default"]("".concat(classPrefix, "-upload__single"), "".concat(classPrefix, "-upload__single-").concat(display));
  var showProgress = showUploadProgress && file && file.status !== "success";
  var inputName = file === null || file === void 0 ? void 0 : file.name;
  var handleRemove = React.useCallback(function (e) {
    onRemove === null || onRemove === void 0 ? void 0 : onRemove({
      e: e,
      file: file,
      index: 0
    });
  }, [file, onRemove]);
  var renderProgress = React.useCallback(function () {
    if ((file === null || file === void 0 ? void 0 : file.status) === "fail") {
      return /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ErrorCircleFilledIcon, null);
    }

    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__single-progress")
    }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, {
      loading: true,
      size: "small"
    }), /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__single-percent")
    }, Math.min((file === null || file === void 0 ? void 0 : file.percent) || 0, 99), "%"));
  }, [classPrefix, file]);
  var renderFilePreviewAsText = React.useCallback(function () {
    if (!inputName) return;
    return /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(classPrefix, "-upload__single-display-text ").concat(classPrefix, "-upload__display-text--margin")
    }, /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(classPrefix, "-upload__single-name")
    }, inputName), showProgress ? renderProgress() : /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CloseCircleFilledIcon, {
      className: "".concat(classPrefix, "-upload__icon-delete"),
      onClick: handleRemove
    }));
  }, [classPrefix, inputName, renderProgress, showProgress, handleRemove]);
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: fileClass
  }, props.children, display === "file" && renderFilePreviewAsText());
};

exports["default"] = SingleFile;
