// tdesign-react/es/upload/themes/image-card.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-744640f5.js';
import React, { useState, Fragment } from 'react';
import { BrowseIcon, DeleteIcon, AddIcon } from 'tdesign-icons-react';
import { Loading } from '../../loading/index.js';
import { Dialog } from '../../dialog/index.js';
import useConfig from '../../_util/useConfig.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import { finishUpload } from '../util.js';
import BooleanRender from '../boolean-render.js';
import '../../_chunks/dep-0e5be0b2.js';
import '../../loading/Loading.js';
import '../../_chunks/dep-08260abc.js';
import 'classnames';
import '../../_util/dom.js';
import 'raf';
import '../../_chunks/dep-60c49e74.js';
import '../../_chunks/dep-12690e1d.js';
import '../../_chunks/dep-21baa91f.js';
import '../../_chunks/dep-4705ff31.js';
import '../../_chunks/dep-a1fd625f.js';
import '../../_chunks/dep-12718f01.js';
import '../../_util/easing.js';
import '../../common/Portal.js';
import 'react-dom';
import '../../_chunks/dep-e3ed7551.js';
import '../../_chunks/dep-ac0cc4dd.js';
import '../../_chunks/dep-e3ab6b0d.js';
import '../../_chunks/dep-cc4f2bae.js';
import '../../_chunks/dep-531703d1.js';
import '../../_chunks/dep-38424935.js';
import '../../_chunks/dep-ad150068.js';
import '../../_chunks/dep-d5f141ac.js';
import '../../_chunks/dep-1b27e5c2.js';
import '../../_chunks/dep-6963df41.js';
import '../../_chunks/dep-8fb1642c.js';
import '../../_chunks/dep-bcb59ae2.js';
import '../../_chunks/dep-09c318fb.js';
import '../../_chunks/dep-048c506f.js';
import '../../_chunks/dep-febf8cf9.js';
import '../../_chunks/dep-074143e4.js';
import '../../_chunks/dep-f5ab8a5c.js';
import '../../_chunks/dep-67009882.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../config-provider/zh_CN_config.js';
import '../../loading/gradient.js';
import '../../_common/js/loading/circle-adapter.js';
import '../../_common/js/utils/set-style.js';
import '../../_common/js/utils/helper.js';
import '../../_chunks/dep-8ff4a741.js';
import '../../loading/defaultProps.js';
import '../../loading/plugin.js';
import './style/css.js';
import '../../loading/type.js';
import '../../dialog/Dialog.js';
import '../../button/index.js';
import '../../button/Button.js';
import '../../_util/useRipple.js';
import '../../_util/useAnimation.js';
import '../../_util/setStyle.js';
import '../../button/defaultProps.js';
import '../../button/type.js';
import '../../_util/noop.js';
import '../../dialog/RenderDialog.js';
import 'react-transition-group';
import '../../_util/useLayoutEffect.js';
import '../../_util/useDialogEsc.js';
import '../../_chunks/dep-bd2639ad.js';
import '../../dialog/defaultProps.js';
import '../../_util/useSetState.js';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../dialog/plugin.js';
import '../../dialog/type.js';

var ImageCard = function ImageCard(props) {
  var files = props.files,
      _props$multiple = props.multiple,
      multiple = _props$multiple === void 0 ? false : _props$multiple,
      _props$max = props.max,
      max = _props$max === void 0 ? 0 : _props$max,
      onRemove = props.onRemove,
      showUploadProgress = props.showUploadProgress;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useLocaleReceiver = useLocaleReceiver("upload"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      showImg = _useState2[0],
      setShowImg = _useState2[1];

  var _useState3 = useState(),
      _useState4 = _slicedToArray(_useState3, 2),
      imgURL = _useState4[0],
      setImgURL = _useState4[1];

  var preview = function preview(file) {
    setShowImg(true);
    setImgURL(file.url);
  };

  var showTrigger = React.useMemo(function () {
    if (multiple) {
      return !max || files.length < max;
    }

    return !(files && files[0]);
  }, [files, max, multiple]);
  return /* @__PURE__ */React.createElement(Fragment, null, /* @__PURE__ */React.createElement(Dialog, {
    visible: showImg,
    showOverlay: true,
    width: "auto",
    top: "10%",
    className: "".concat(classPrefix, "-upload__dialog"),
    footer: false,
    header: false,
    onClose: function onClose() {
      setShowImg(false);
      setImgURL(null);
    }
  }, /* @__PURE__ */React.createElement("p", {
    className: "".concat(classPrefix, "-dialog__dialog-body-img-box")
  }, /* @__PURE__ */React.createElement("img", {
    src: imgURL,
    alt: ""
  }))), /* @__PURE__ */React.createElement("ul", {
    className: "".concat(classPrefix, "-upload__card")
  }, /* @__PURE__ */React.createElement(BooleanRender, {
    boolExpression: !!files
  }, files.map(function (file, index) {
    return /* @__PURE__ */React.createElement("li", {
      className: "".concat(classPrefix, "-upload__card-item ").concat(classPrefix, "-is--background"),
      key: index
    }, /* @__PURE__ */React.createElement(BooleanRender, {
      falseRender: function falseRender() {
        return /* @__PURE__ */React.createElement("div", {
          className: "".concat(classPrefix, "-upload__card-content ").concat(classPrefix, "-upload__card-box")
        }, /* @__PURE__ */React.createElement("img", {
          className: "".concat(classPrefix, "-upload__card-image"),
          src: file.url
        }), /* @__PURE__ */React.createElement("div", {
          className: "".concat(classPrefix, "-upload__card-mask")
        }, /* @__PURE__ */React.createElement("span", {
          className: "".concat(classPrefix, "-upload__card-mask-item"),
          onClick: function onClick(e) {
            return e.stopPropagation();
          }
        }, /* @__PURE__ */React.createElement(BrowseIcon, {
          onClick: function onClick() {
            preview(file);
          }
        })), /* @__PURE__ */React.createElement("span", {
          className: "".concat(classPrefix, "-upload__card-mask-item-divider")
        }), /* @__PURE__ */React.createElement("span", {
          className: "".concat(classPrefix, "-upload__card-mask-item"),
          onClick: function onClick(e) {
            return e.stopPropagation();
          }
        }, /* @__PURE__ */React.createElement(DeleteIcon, {
          onClick: function onClick(e) {
            onRemove === null || onRemove === void 0 ? void 0 : onRemove({
              e: e,
              file: file,
              index: index
            });
          }
        }))));
      },
      boolExpression: showUploadProgress && !finishUpload(file.status)
    }, /* @__PURE__ */React.createElement("div", {
      className: "".concat(classPrefix, "-upload__card-container ").concat(classPrefix, "-upload__card-box")
    }, /* @__PURE__ */React.createElement(Loading, {
      loading: true,
      size: "medium"
    }), /* @__PURE__ */React.createElement("p", null, t(locale.progress.uploadingText), " ", Math.min(files[index].percent, 99), "%"))));
  })), /* @__PURE__ */React.createElement(BooleanRender, {
    boolExpression: showTrigger
  }, /* @__PURE__ */React.createElement("li", {
    className: "".concat(classPrefix, "-upload__card-item ").concat(classPrefix, "-is--background"),
    onClick: props.onTrigger
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-upload__card-container ").concat(classPrefix, "-upload__card-box")
  }, /* @__PURE__ */React.createElement(AddIcon, null), /* @__PURE__ */React.createElement("p", {
    className: "".concat(classPrefix, "-size-s")
  }, t(locale.triggerUploadText.image)))))));
};

export { ImageCard as default };
