// tdesign-react/es/common/Jumper.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import React from 'react';
import { ChevronLeftIcon, RoundIcon, ChevronRightIcon } from 'tdesign-icons-react';
import useConfig from '../_util/useConfig.js';
import { Button } from '../button/index.js';
import '../_chunks/dep-e3ed7551.js';
import '../_chunks/dep-08260abc.js';
import '../_chunks/dep-12690e1d.js';
import '../_chunks/dep-ac0cc4dd.js';
import '../_chunks/dep-e3ab6b0d.js';
import '../_chunks/dep-cc4f2bae.js';
import '../_chunks/dep-531703d1.js';
import '../_chunks/dep-21baa91f.js';
import '../_chunks/dep-4705ff31.js';
import '../_chunks/dep-38424935.js';
import '../_chunks/dep-ad150068.js';
import '../_chunks/dep-d5f141ac.js';
import '../_chunks/dep-1b27e5c2.js';
import '../_chunks/dep-6963df41.js';
import '../_chunks/dep-8fb1642c.js';
import '../_chunks/dep-bcb59ae2.js';
import '../_chunks/dep-12718f01.js';
import '../_chunks/dep-09c318fb.js';
import '../_chunks/dep-048c506f.js';
import '../_chunks/dep-febf8cf9.js';
import '../_chunks/dep-074143e4.js';
import '../_chunks/dep-a1fd625f.js';
import '../_chunks/dep-f5ab8a5c.js';
import '../_chunks/dep-67009882.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../config-provider/zh_CN_config.js';
import '../button/Button.js';
import '../_chunks/dep-8ff4a741.js';
import 'classnames';
import '../_util/useRipple.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-60c49e74.js';
import '../_util/easing.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../_chunks/dep-744640f5.js';
import '../_chunks/dep-0e5be0b2.js';
import './Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import './style/css.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../button/type.js';

function Jumper(props) {
  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var prevTitle = props.prevTitle,
      currentTitle = props.currentTitle,
      nextTitle = props.nextTitle,
      onJumperClick = props.onJumperClick,
      _props$size = props.size,
      size = _props$size === void 0 ? "small" : _props$size;
  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-jumper")
  }, /* @__PURE__ */React.createElement(Button, {
    title: prevTitle,
    variant: "text",
    size: size,
    shape: "square",
    onClick: function onClick() {
      return onJumperClick(-1);
    },
    icon: /* @__PURE__ */React.createElement(ChevronLeftIcon, null),
    className: "".concat(classPrefix, "-jumper__btn")
  }), /* @__PURE__ */React.createElement(Button, {
    title: currentTitle,
    variant: "text",
    size: size,
    shape: "square",
    onClick: function onClick() {
      return onJumperClick(0);
    },
    icon: /* @__PURE__ */React.createElement(RoundIcon, null),
    className: "".concat(classPrefix, "-jumper__btn")
  }), /* @__PURE__ */React.createElement(Button, {
    title: nextTitle,
    variant: "text",
    size: size,
    shape: "square",
    onClick: function onClick() {
      return onJumperClick(1);
    },
    icon: /* @__PURE__ */React.createElement(ChevronRightIcon, null),
    className: "".concat(classPrefix, "-jumper__btn")
  }));
}

export { Jumper as default };
