// tdesign-react/lib/date-picker/panel/DateRangePickerPanel.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-a1655de1.js');
var slicedToArray = require('../../_chunks/dep-3cf9becb.js');
var React = require('react');
var classNames = require('classnames');
var dayjs = require('dayjs');
var _util_useConfig = require('../../_util/useConfig.js');
var datePicker_panel_PanelContent = require('./PanelContent.js');
var datePicker_panel_ExtraContent = require('./ExtraContent.js');
var datePicker_panel_useTableData = require('./useTableData.js');
var datePicker_hooks_useDisableDate = require('../hooks/useDisableDate.js');
require('../../_chunks/dep-19841e1d.js');
require('../../_chunks/dep-38a46f48.js');
require('../../_chunks/dep-8a149e9f.js');
require('../../_chunks/dep-3b69722d.js');
require('../../_chunks/dep-fde4d28a.js');
require('../../_chunks/dep-e266830c.js');
require('../../_chunks/dep-fcf32c81.js');
require('../../_chunks/dep-3405bda4.js');
require('../../_chunks/dep-15a4424e.js');
require('../../_chunks/dep-c40d75ac.js');
require('../../_chunks/dep-910dcfa3.js');
require('../../_chunks/dep-633a5671.js');
require('../../_chunks/dep-9c339ffc.js');
require('../../_chunks/dep-b1a67107.js');
require('../../_chunks/dep-ab43c961.js');
require('../../_chunks/dep-9cacaf01.js');
require('../../_chunks/dep-c72c7229.js');
require('../../_chunks/dep-8e0a771d.js');
require('../../_chunks/dep-c90cc28f.js');
require('../../_chunks/dep-542cc308.js');
require('../../_chunks/dep-c6637e11.js');
require('../../_common/js/global-config/locale/zh_CN.js');
require('../../config-provider/zh_CN_config.js');
require('../base/Header.js');
require('../../_chunks/dep-f6a4bdd5.js');
require('../../locale/LocalReceiver.js');
require('../../select/index.js');
require('../../select/base/Select.js');
require('../../_chunks/dep-23535a52.js');
require('../../_chunks/dep-3181fd9f.js');
require('../../_chunks/dep-8a3a8f07.js');
require('../../_chunks/dep-cb0f33ac.js');
require('../../_chunks/dep-86abec96.js');
require('../../tag/index.js');
require('../../tag/Tag.js');
require('tdesign-icons-react');
require('../../_util/noop.js');
require('../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../tag/CheckTag.js');
require('../../hooks/useControlled.js');
require('../../_chunks/dep-f3090217.js');
require('../../_chunks/dep-d4a0b3f0.js');
require('../../_chunks/dep-6dfb2577.js');
require('../../tag/defaultProps.js');
require('../../select/util/helper.js');
require('../../select/base/OptionGroup.js');
require('../../select/defaultProps.js');
require('../../select/base/Option.js');
require('../../_chunks/dep-bc80a1a5.js');
require('../../_chunks/dep-1d819290.js');
require('../../_util/useRipple.js');
require('../../_util/useAnimation.js');
require('../../_util/setStyle.js');
require('../../_util/dom.js');
require('raf');
require('../../_util/easing.js');
require('../../common/FakeArrow.js');
require('../../loading/index.js');
require('../../loading/Loading.js');
require('../../common/Portal.js');
require('react-dom');
require('../../loading/gradient.js');
require('../../_common/js/loading/circle-adapter.js');
require('../../_common/js/utils/set-style.js');
require('../../_common/js/utils/helper.js');
require('../../loading/defaultProps.js');
require('../../loading/plugin.js');
require('../../select-input/index.js');
require('../../select-input/SelectInput.js');
require('../../popup/index.js');
require('../../popup/Popup.js');
require('react-transition-group');
require('react-popper');
require('../../_util/composeRefs.js');
require('../../popup/hooks/useTriggerProps.js');
require('../../_util/useClickOutside.js');
require('../../popup/utils/getTransitionParams.js');
require('../../_util/useMutationObserver.js');
require('../../_chunks/dep-85b3eaa4.js');
require('../../_chunks/dep-e19f22f6.js');
require('../../_util/useWindowSize.js');
require('../../popup/defaultProps.js');
require('../../select-input/useSingle.js');
require('../../_chunks/dep-46e6aa42.js');
require('../../_chunks/dep-c7b74e6e.js');
require('../../_chunks/dep-c3bd0e8c.js');
require('../../_chunks/dep-38856994.js');
require('../../_chunks/dep-ec5420f8.js');
require('../../input/index.js');
require('../../input/Input.js');
require('../../_util/helper.js');
require('../../_chunks/dep-e3db5727.js');
require('../../input/InputGroup.js');
require('../../input/defaultProps.js');
require('../../select-input/useMultiple.js');
require('../../tag-input/index.js');
require('../../tag-input/TagInput.js');
require('../../_util/useDragSorter.js');
require('../../tag-input/useTagScroll.js');
require('../../tag-input/useTagList.js');
require('lodash');
require('../../tag-input/useHover.js');
require('../../tag-input/defaultProps.js');
require('../../select-input/useOverlayStyle.js');
require('../../select-input/defaultProps.js');
require('../../select/base/PopupContent.js');
require('../../common/Jumper.js');
require('../../button/index.js');
require('../../button/Button.js');
require('../../button/defaultProps.js');
require('../base/Table.js');
require('../base/Cell.js');
require('../../_common/js/date-picker/utils-new.js');
require('../../_chunks/dep-e9185c09.js');
require('../../_chunks/dep-d97f1281.js');
require('../../time-picker/panel/TimePickerPanel.js');
require('../../_chunks/dep-486c2bda.js');
require('../../_chunks/dep-837b19c6.js');
require('../../time-picker/hooks/useTimePickerTextConfig.js');
require('../../_common/js/time-picker/const.js');
require('../../_common/js/time-picker/utils.js');
require('../hooks/useFormat.js');
require('../base/Footer.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { defineProperty._defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var DateRangePickerPanel = function DateRangePickerPanel(props) {
  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix,
      globalDatePickerConfig = _useConfig.datePicker;

  var panelName = "".concat(classPrefix, "-date-range-picker__panel");
  var _props$value = props.value,
      value = _props$value === void 0 ? [] : _props$value,
      _props$hoverValue = props.hoverValue,
      hoverValue = _props$hoverValue === void 0 ? [] : _props$hoverValue,
      mode = props.mode,
      format = props.format,
      presets = props.presets,
      enableTimePicker = props.enableTimePicker,
      presetsPlacement = props.presetsPlacement,
      disableDateFromProps = props.disableDate,
      _props$firstDayOfWeek = props.firstDayOfWeek,
      firstDayOfWeek = _props$firstDayOfWeek === void 0 ? globalDatePickerConfig.firstDayOfWeek : _props$firstDayOfWeek,
      isFirstValueSelected = props.isFirstValueSelected,
      style = props.style,
      className = props.className,
      activeIndex = props.activeIndex,
      year = props.year,
      month = props.month,
      _props$timeValue = props.timeValue,
      timeValue = _props$timeValue === void 0 ? [] : _props$timeValue,
      _onClick = props.onClick,
      onConfirmClick = props.onConfirmClick,
      onPresetClick = props.onPresetClick;
  var disableDateOptions = datePicker_hooks_useDisableDate["default"]({
    disableDate: disableDateFromProps,
    mode: mode,
    format: format,
    start: isFirstValueSelected && activeIndex === 1 ? dayjs__default["default"](value[0]).toDate() : void 0,
    end: isFirstValueSelected && activeIndex === 0 ? dayjs__default["default"](value[1]).toDate() : void 0
  });

  var _year = slicedToArray._slicedToArray(year, 2),
      startYear = _year[0],
      endYear = _year[1];

  var _month = slicedToArray._slicedToArray(month, 2),
      startMonth = _month[0],
      endMonth = _month[1];

  var startTableData = datePicker_panel_useTableData["default"](_objectSpread({
    isRange: true,
    start: value[0] ? dayjs__default["default"](value[0]).toDate() : void 0,
    end: value[1] ? dayjs__default["default"](value[1]).toDate() : void 0,
    hoverStart: hoverValue[0] ? dayjs__default["default"](hoverValue[0]).toDate() : void 0,
    hoverEnd: hoverValue[1] ? dayjs__default["default"](hoverValue[1]).toDate() : void 0,
    year: startYear,
    month: startMonth,
    mode: mode,
    firstDayOfWeek: firstDayOfWeek
  }, disableDateOptions));
  var endTableData = datePicker_panel_useTableData["default"](_objectSpread({
    isRange: true,
    start: value[0] ? dayjs__default["default"](value[0]).toDate() : void 0,
    end: value[1] ? dayjs__default["default"](value[1]).toDate() : void 0,
    hoverStart: hoverValue[0] ? dayjs__default["default"](hoverValue[0]).toDate() : void 0,
    hoverEnd: hoverValue[1] ? dayjs__default["default"](hoverValue[1]).toDate() : void 0,
    year: endYear,
    month: endMonth,
    mode: mode,
    firstDayOfWeek: firstDayOfWeek
  }, disableDateOptions));
  var panelContentProps = {
    mode: mode,
    format: format,
    firstDayOfWeek: firstDayOfWeek,
    enableTimePicker: props.enableTimePicker,
    timePickerProps: props.timePickerProps,
    onMonthChange: props.onMonthChange,
    onYearChange: props.onYearChange,
    onJumperClick: props.onJumperClick,
    onCellClick: props.onCellClick,
    onCellMouseEnter: props.onCellMouseEnter,
    onCellMouseLeave: props.onCellMouseLeave,
    onTimePickerChange: props.onTimePickerChange
  };
  return /* @__PURE__ */React__default["default"].createElement("div", {
    style: style,
    className: classNames__default["default"](panelName, className, defineProperty._defineProperty({}, "".concat(panelName, "--direction-row"), ["left", "right"].includes(presetsPlacement))),
    onClick: function onClick(e) {
      return _onClick === null || _onClick === void 0 ? void 0 : _onClick({
        e: e
      });
    }
  }, ["top", "left"].includes(presetsPlacement) ? /* @__PURE__ */React__default["default"].createElement(datePicker_panel_ExtraContent["default"], {
    presets: presets,
    selectedValue: value[activeIndex],
    enableTimePicker: enableTimePicker,
    onPresetClick: onPresetClick,
    onConfirmClick: onConfirmClick,
    presetsPlacement: presetsPlacement
  }) : null, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(panelName, "--content-wrapper")
  }, !enableTimePicker ? [/* @__PURE__ */React__default["default"].createElement(datePicker_panel_PanelContent["default"], _objectSpread({
    key: "startPanel",
    partial: "start",
    year: startYear,
    month: startMonth,
    timeValue: timeValue[0],
    tableData: startTableData
  }, panelContentProps)), /* @__PURE__ */React__default["default"].createElement(datePicker_panel_PanelContent["default"], _objectSpread({
    key: "endPanel",
    partial: "end",
    year: endYear,
    month: endMonth,
    timeValue: timeValue[1],
    tableData: endTableData
  }, panelContentProps))] : /* @__PURE__ */React__default["default"].createElement(datePicker_panel_PanelContent["default"], _objectSpread({
    key: "start",
    partial: activeIndex ? "end" : "start",
    year: activeIndex ? endYear : startYear,
    month: activeIndex ? endMonth : startMonth,
    timeValue: activeIndex ? timeValue[1] : timeValue[0],
    tableData: activeIndex ? endTableData : startTableData
  }, panelContentProps))), ["bottom", "right"].includes(presetsPlacement) ? /* @__PURE__ */React__default["default"].createElement(datePicker_panel_ExtraContent["default"], {
    presets: presets,
    selectedValue: value[activeIndex],
    enableTimePicker: enableTimePicker,
    onPresetClick: onPresetClick,
    onConfirmClick: onConfirmClick,
    presetsPlacement: presetsPlacement
  }) : null);
};

DateRangePickerPanel.displayName = "DateRangePickerPanel";
DateRangePickerPanel.defaultProps = {
  mode: "date",
  enableTimePicker: false,
  presetsPlacement: "bottom"
};

exports["default"] = DateRangePickerPanel;
