// tdesign-react/lib/date-picker/panel/useTableData.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var slicedToArray = require('../../_chunks/dep-3cf9becb.js');
var locale_LocalReceiver = require('../../locale/LocalReceiver.js');
var _common_js_datePicker_utilsNew = require('../../_common/js/date-picker/utils-new.js');
require('../../_chunks/dep-19841e1d.js');
require('../../_chunks/dep-a1655de1.js');
require('react');
require('../../_chunks/dep-38a46f48.js');
require('../../_chunks/dep-8a149e9f.js');
require('../../_chunks/dep-3b69722d.js');
require('../../_chunks/dep-fde4d28a.js');
require('../../_chunks/dep-e266830c.js');
require('../../_chunks/dep-fcf32c81.js');
require('../../_chunks/dep-3405bda4.js');
require('../../_chunks/dep-15a4424e.js');
require('../../_chunks/dep-c40d75ac.js');
require('../../_chunks/dep-910dcfa3.js');
require('../../_chunks/dep-633a5671.js');
require('../../_chunks/dep-9c339ffc.js');
require('../../_chunks/dep-b1a67107.js');
require('../../_chunks/dep-ab43c961.js');
require('../../_chunks/dep-9cacaf01.js');
require('../../_chunks/dep-c72c7229.js');
require('../../_chunks/dep-8e0a771d.js');
require('../../_chunks/dep-c90cc28f.js');
require('../../_chunks/dep-542cc308.js');
require('../../_chunks/dep-c6637e11.js');
require('../../_common/js/global-config/locale/zh_CN.js');
require('../../config-provider/zh_CN_config.js');
require('../../_chunks/dep-e9185c09.js');
require('../../_chunks/dep-6dfb2577.js');
require('../../_chunks/dep-d97f1281.js');
require('../../_chunks/dep-e19f22f6.js');
require('../../_chunks/dep-cb0f33ac.js');
require('dayjs');

function useTableData(props) {
  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("datePicker"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var monthLocal = t(local.months);
  var start = props.start,
      end = props.end,
      hoverStart = props.hoverStart,
      hoverEnd = props.hoverEnd,
      year = props.year,
      month = props.month,
      mode = props.mode,
      firstDayOfWeek = props.firstDayOfWeek,
      disableDate = props.disableDate,
      minDate = props.minDate,
      maxDate = props.maxDate,
      isRange = props.isRange;
  var data = [];
  var options = {
    minDate: minDate,
    maxDate: maxDate,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    monthLocal: monthLocal
  };

  if (mode === "date") {
    data = _common_js_datePicker_utilsNew.getWeeks({
      year: year,
      month: month
    }, options);
  } else if (mode === "month") {
    data = _common_js_datePicker_utilsNew.getMonths(year, options);
  } else if (mode === "year") {
    data = _common_js_datePicker_utilsNew.getYears(year, options);
  }

  return _common_js_datePicker_utilsNew.flagActive(data, {
    start: start,
    end: end,
    hoverStart: hoverStart,
    hoverEnd: hoverEnd,
    type: mode,
    isRange: isRange
  });
}

exports["default"] = useTableData;
