// tdesign-react/lib/upload/themes/flow-list/img-list.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-a1655de1.js');
var slicedToArray = require('../../../_chunks/dep-3cf9becb.js');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var React = require('react');
var _util_useConfig = require('../../../_util/useConfig.js');
var locale_LocalReceiver = require('../../../locale/LocalReceiver.js');
var upload_util = require('../../util.js');
var loading_index = require('../../../loading/index.js');
require('../../../_chunks/dep-19841e1d.js');
require('../../../_chunks/dep-38a46f48.js');
require('../../../_chunks/dep-8a149e9f.js');
require('../../../_chunks/dep-3b69722d.js');
require('../../../_chunks/dep-fde4d28a.js');
require('../../../_chunks/dep-e266830c.js');
require('../../../_chunks/dep-fcf32c81.js');
require('../../../_chunks/dep-3405bda4.js');
require('../../../_chunks/dep-15a4424e.js');
require('../../../_chunks/dep-c40d75ac.js');
require('../../../_chunks/dep-910dcfa3.js');
require('../../../_chunks/dep-633a5671.js');
require('../../../_chunks/dep-9c339ffc.js');
require('../../../_chunks/dep-b1a67107.js');
require('../../../_chunks/dep-ab43c961.js');
require('../../../_chunks/dep-9cacaf01.js');
require('../../../_chunks/dep-c72c7229.js');
require('../../../_chunks/dep-8e0a771d.js');
require('../../../_chunks/dep-c90cc28f.js');
require('../../../_chunks/dep-542cc308.js');
require('../../../_chunks/dep-c6637e11.js');
require('../../../_common/js/global-config/locale/zh_CN.js');
require('../../../config-provider/zh_CN_config.js');
require('../../../loading/Loading.js');
require('../../../_util/dom.js');
require('raf');
require('../../../_chunks/dep-1d819290.js');
require('../../../_util/easing.js');
require('../../../common/Portal.js');
require('react-dom');
require('../../../loading/gradient.js');
require('../../../_common/js/loading/circle-adapter.js');
require('../../../_common/js/utils/set-style.js');
require('../../../_common/js/utils/helper.js');
require('../../../_chunks/dep-23535a52.js');
require('../../../loading/defaultProps.js');
require('../../../loading/plugin.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var ImgList = function ImgList(props) {
  var showInitial = props.showInitial,
      listFiles = props.listFiles,
      renderDragger = props.renderDragger,
      onImgPreview = props.onImgPreview,
      remove = props.remove;

  var _useConfig = _util_useConfig["default"](),
      prefix = _useConfig.classPrefix;

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("upload"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var UPLOAD_NAME = "".concat(prefix, "-upload");
  var UPLOAD_NAME_CARD = "".concat(UPLOAD_NAME, "__card");
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(UPLOAD_NAME, "__flow-card-area")
  }, showInitial && renderDragger(), !!listFiles.length && /* @__PURE__ */React__default["default"].createElement("ul", {
    className: "".concat(UPLOAD_NAME_CARD, " clearfix")
  }, listFiles.map(function (file, index) {
    var _classNames;

    return /* @__PURE__ */React__default["default"].createElement("li", {
      className: "".concat(UPLOAD_NAME_CARD, "-item"),
      key: index
    }, /* @__PURE__ */React__default["default"].createElement("div", {
      className: classNames__default["default"]((_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(UPLOAD_NAME_CARD, "-content"), true), defineProperty._defineProperty(_classNames, "".concat(UPLOAD_NAME_CARD, "-content-border"), file.status !== "waiting"), _classNames))
    }, file.status === "fail" && /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-status-wrap")
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.ErrorCircleFilledIcon, null), /* @__PURE__ */React__default["default"].createElement("p", null, t(locale.progress.failText))), file.status === "progress" && /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-status-wrap")
    }, /* @__PURE__ */React__default["default"].createElement(loading_index.Loading, null), /* @__PURE__ */React__default["default"].createElement("p", null, t(locale.progress.uploadingText), " ", Math.min(file.percent, 99))), (["waiting", "success"].includes(file.status) || !file.status && file.url) && /* @__PURE__ */React__default["default"].createElement("img", {
      className: "".concat(UPLOAD_NAME_CARD, "-image"),
      src: file.url || "//tdesign.gtimg.com/tdesign-default-img.png",
      alt: ""
    }), /* @__PURE__ */React__default["default"].createElement("div", {
      className: "".concat(UPLOAD_NAME_CARD, "-mask")
    }, file.url && /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "__mask__item")
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.BrowseIcon, {
      onClick: function onClick(e) {
        return onImgPreview(file, e);
      }
    }), /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "__mask__item-divider")
    })), /* @__PURE__ */React__default["default"].createElement("span", {
      className: "".concat(UPLOAD_NAME_CARD, "-mask__item"),
      onClick: function onClick(e) {
        return remove({
          e: e,
          index: index,
          file: file
        });
      }
    }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.DeleteIcon, null)))), /* @__PURE__ */React__default["default"].createElement("p", {
      className: "".concat(UPLOAD_NAME_CARD, "-name")
    }, upload_util.abridgeName(file.name)));
  })));
};

exports["default"] = ImgList;
