// tdesign-react/lib/upload/themes/flow-list/index.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-a1655de1.js');
var slicedToArray = require('../../../_chunks/dep-3cf9becb.js');
var React = require('react');
var classNames = require('classnames');
var _util_useConfig = require('../../../_util/useConfig.js');
var locale_LocalReceiver = require('../../../locale/LocalReceiver.js');
var button_index = require('../../../button/index.js');
var upload_themes_flowList_imgList = require('./img-list.js');
var upload_themes_flowList_fileList = require('./file-list.js');
var upload_booleanRender = require('../../boolean-render.js');
require('../../../_chunks/dep-19841e1d.js');
require('../../../_chunks/dep-38a46f48.js');
require('../../../_chunks/dep-8a149e9f.js');
require('../../../_chunks/dep-3b69722d.js');
require('../../../_chunks/dep-fde4d28a.js');
require('../../../_chunks/dep-e266830c.js');
require('../../../_chunks/dep-fcf32c81.js');
require('../../../_chunks/dep-3405bda4.js');
require('../../../_chunks/dep-15a4424e.js');
require('../../../_chunks/dep-c40d75ac.js');
require('../../../_chunks/dep-910dcfa3.js');
require('../../../_chunks/dep-633a5671.js');
require('../../../_chunks/dep-9c339ffc.js');
require('../../../_chunks/dep-b1a67107.js');
require('../../../_chunks/dep-ab43c961.js');
require('../../../_chunks/dep-9cacaf01.js');
require('../../../_chunks/dep-c72c7229.js');
require('../../../_chunks/dep-8e0a771d.js');
require('../../../_chunks/dep-c90cc28f.js');
require('../../../_chunks/dep-542cc308.js');
require('../../../_chunks/dep-c6637e11.js');
require('../../../_common/js/global-config/locale/zh_CN.js');
require('../../../config-provider/zh_CN_config.js');
require('../../../button/Button.js');
require('../../../_chunks/dep-23535a52.js');
require('../../../_util/useRipple.js');
require('../../../_util/useAnimation.js');
require('../../../_util/setStyle.js');
require('../../../_util/dom.js');
require('raf');
require('../../../_chunks/dep-1d819290.js');
require('../../../_util/easing.js');
require('../../../loading/index.js');
require('../../../loading/Loading.js');
require('../../../common/Portal.js');
require('react-dom');
require('../../../loading/gradient.js');
require('../../../_common/js/loading/circle-adapter.js');
require('../../../_common/js/utils/set-style.js');
require('../../../_common/js/utils/helper.js');
require('../../../loading/defaultProps.js');
require('../../../loading/plugin.js');
require('../../../button/defaultProps.js');
require('tdesign-icons-react');
require('../../util.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Index = function Index(props) {
  var _classNames;

  var listFiles = props.files,
      _props$toUploadFiles = props.toUploadFiles,
      toUploadFiles = _props$toUploadFiles === void 0 ? [] : _props$toUploadFiles,
      _props$showUploadProg = props.showUploadProgress,
      showUploadProgress = _props$showUploadProg === void 0 ? false : _props$showUploadProg,
      placeholder = props.placeholder,
      display = props.display,
      onImgPreview = props.onImgPreview,
      remove = props.remove,
      upload = props.upload,
      cancel = props.cancel,
      onChange = props.onChange,
      onDragenter = props.onDragenter,
      onDragleave = props.onDragleave,
      _props$children = props.children,
      children = _props$children === void 0 ? null : _props$children;
  var target = React__default["default"].useRef();

  var _useConfig = _util_useConfig["default"](),
      prefix = _useConfig.classPrefix;

  var _useLocaleReceiver = locale_LocalReceiver.useLocaleReceiver("upload"),
      _useLocaleReceiver2 = slicedToArray._slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var UPLOAD_NAME = "".concat(prefix, "-upload");

  var _useState = React.useState(false),
      _useState2 = slicedToArray._slicedToArray(_useState, 2),
      dragActive = _useState2[0],
      setDragActive = _useState2[1];

  var showInitial = !listFiles.length;
  var failedList = toUploadFiles.filter(function (file) {
    return file.status === "fail";
  });
  var isUploading = toUploadFiles.filter(function (file) {
    return file.status === "progress";
  }).length > 0;
  var allowUpload = toUploadFiles.length > 0 && !isUploading;
  var progress = locale.progress,
      triggerUploadText = locale.triggerUploadText;
  var uploadText = failedList.length ? t(triggerUploadText.reupload) : t(triggerUploadText.normal);
  if (isUploading) uploadText = t(progress.uploadingText);

  var handleDrop = function handleDrop(event) {
    event.preventDefault();
    onChange === null || onChange === void 0 ? void 0 : onChange(event.dataTransfer.files);
    onDragleave === null || onDragleave === void 0 ? void 0 : onDragleave(event);
    setDragActive(false);
  };

  var handleDragenter = function handleDragenter(event) {
    event.preventDefault();
    onDragenter === null || onDragenter === void 0 ? void 0 : onDragenter(event);
    setDragActive(true);
  };

  var handleDragleave = function handleDragleave(event) {
    if (event.target !== target.current) return;
    event.preventDefault();
    onDragleave === null || onDragleave === void 0 ? void 0 : onDragleave(event);
    setDragActive(false);
  };

  var handleDragover = function handleDragover(event) {
    event.preventDefault();
  };

  var onInternalImgPreview = React.useCallback(function (file, event) {
    if (!onImgPreview) return;
    event.preventDefault();
    onImgPreview(file, event);
  }, [onImgPreview]);

  var renderDragger = function renderDragger() {
    return /* @__PURE__ */React__default["default"].createElement("div", {
      ref: target,
      className: "".concat(UPLOAD_NAME, "__flow-empty"),
      onDrop: handleDrop,
      onDragEnter: handleDragenter,
      onDragOver: handleDragover,
      onDragLeave: handleDragleave
    }, dragActive ? t(locale.dragger.dragDropText) : t(locale.dragger.clickAndDragText));
  };

  var wrapperClassNames = classNames__default["default"]((_classNames = {}, defineProperty._defineProperty(_classNames, "".concat(UPLOAD_NAME, "__flow"), true), defineProperty._defineProperty(_classNames, "".concat(UPLOAD_NAME, "__flow-").concat(display), true), _classNames));
  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: wrapperClassNames
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(UPLOAD_NAME, "__flow-op")
  }, children, /* @__PURE__ */React__default["default"].createElement("small", {
    className: "".concat(UPLOAD_NAME, "__small ").concat(UPLOAD_NAME, "__flow-placeholder")
  }, placeholder)), /* @__PURE__ */React__default["default"].createElement(upload_booleanRender["default"], {
    boolExpression: display === "file-flow"
  }, /* @__PURE__ */React__default["default"].createElement(upload_themes_flowList_fileList["default"], {
    listFiles: listFiles,
    showInitial: showInitial,
    renderDragger: renderDragger,
    showUploadProgress: showUploadProgress,
    remove: remove
  })), /* @__PURE__ */React__default["default"].createElement(upload_booleanRender["default"], {
    boolExpression: display === "image-flow"
  }, /* @__PURE__ */React__default["default"].createElement(upload_themes_flowList_imgList["default"], {
    listFiles: listFiles,
    showInitial: showInitial,
    renderDragger: renderDragger,
    onImgPreview: onInternalImgPreview,
    remove: remove
  })), /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(UPLOAD_NAME, "__flow-bottom")
  }, /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    theme: "default",
    onClick: cancel
  }, t(locale.cancelUploadText)), /* @__PURE__ */React__default["default"].createElement(button_index.Button, {
    disabled: !allowUpload,
    theme: "primary",
    onClick: function onClick(e) {
      return upload(toUploadFiles, e);
    }
  }, uploadText)));
};

exports["default"] = Index;
