// tdesign-react/es/upload/dragger.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../_chunks/dep-744640f5.js';
import React, { useState } from 'react';
import classNames from 'classnames';
import useConfig from '../_util/useConfig.js';
import { useLocaleReceiver } from '../locale/LocalReceiver.js';
import DraggerProgress from './themes/dragger-progress.js';
import '../_chunks/dep-0e5be0b2.js';
import '../_chunks/dep-e3ed7551.js';
import '../_chunks/dep-08260abc.js';
import '../_chunks/dep-12690e1d.js';
import '../_chunks/dep-ac0cc4dd.js';
import '../_chunks/dep-e3ab6b0d.js';
import '../_chunks/dep-cc4f2bae.js';
import '../_chunks/dep-531703d1.js';
import '../_chunks/dep-21baa91f.js';
import '../_chunks/dep-4705ff31.js';
import '../_chunks/dep-38424935.js';
import '../_chunks/dep-ad150068.js';
import '../_chunks/dep-d5f141ac.js';
import '../_chunks/dep-1b27e5c2.js';
import '../_chunks/dep-6963df41.js';
import '../_chunks/dep-8fb1642c.js';
import '../_chunks/dep-bcb59ae2.js';
import '../_chunks/dep-12718f01.js';
import '../_chunks/dep-09c318fb.js';
import '../_chunks/dep-048c506f.js';
import '../_chunks/dep-febf8cf9.js';
import '../_chunks/dep-074143e4.js';
import '../_chunks/dep-a1fd625f.js';
import '../_chunks/dep-f5ab8a5c.js';
import '../_chunks/dep-67009882.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../config-provider/zh_CN_config.js';
import '../config-provider/index.js';
import '../config-provider/ConfigProvider.js';
import 'tdesign-icons-react';
import '../button/index.js';
import '../button/Button.js';
import '../_chunks/dep-8ff4a741.js';
import '../_util/useRipple.js';
import '../_util/useAnimation.js';
import '../_util/setStyle.js';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-60c49e74.js';
import '../_util/easing.js';
import '../loading/index.js';
import '../loading/Loading.js';
import '../common/Portal.js';
import 'react-dom';
import '../loading/gradient.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '../loading/defaultProps.js';
import '../loading/plugin.js';
import './style/css.js';
import '../loading/type.js';
import '../button/defaultProps.js';
import '../button/type.js';
import './util.js';

var Dragger = function Dragger(props) {
  var file = props.file,
      display = props.display,
      _onUpload = props.onUpload,
      onRemove = props.onRemove,
      customDraggerRender = props.customDraggerRender,
      onCancel = props.onCancel;

  var _useConfig = useConfig(),
      classPrefix = _useConfig.classPrefix;

  var _useLocaleReceiver = useLocaleReceiver("upload"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      dragActive = _useState2[0],
      setDragActive = _useState2[1];

  var target = React.useRef();
  var _locale$dragger = locale.dragger,
      draggingText = _locale$dragger.draggingText,
      dragDropText = _locale$dragger.dragDropText;
  var dragTriggerText = t(locale.triggerUploadText.normal);
  var classes = classNames("".concat(classPrefix, "-upload__dragger"), !file ? "".concat(classPrefix, "-upload__dragger-center") : "", (file === null || file === void 0 ? void 0 : file.status) === "fail" ? "".concat(classPrefix, "-upload__dragger-error") : "");
  var defaultDragElement = React.useMemo(function () {
    var unActiveElement = /* @__PURE__ */React.createElement("div", null, /* @__PURE__ */React.createElement("span", {
      className: "".concat(classPrefix, "-upload--highlight")
    }, dragTriggerText), /* @__PURE__ */React.createElement("span", null, "\xA0\xA0/\xA0\xA0", t(draggingText)));
    var activeElement = /* @__PURE__ */React.createElement("div", null, t(dragDropText));
    return dragActive ? activeElement : unActiveElement;
  }, [classPrefix, dragActive, dragDropText, dragTriggerText, draggingText, t]);
  var dragElement = React.useMemo(function () {
    var content;

    if (file && display !== "custom") {
      content = /* @__PURE__ */React.createElement(DraggerProgress, {
        onRemove: onRemove,
        display: display,
        onTrigger: props.onTrigger,
        file: file,
        onUpload: function onUpload() {
          _onUpload === null || _onUpload === void 0 ? void 0 : _onUpload(file);
        },
        onCancel: onCancel
      });
    } else {
      content = /* @__PURE__ */React.createElement("div", {
        className: "".concat(classPrefix, "-upload__trigger"),
        onClick: props.onTrigger
      }, (customDraggerRender === null || customDraggerRender === void 0 ? void 0 : customDraggerRender({
        dragActive: dragActive
      })) || defaultDragElement);
    }

    return content;
  }, [file, display, onRemove, props.onTrigger, onCancel, _onUpload, classPrefix, customDraggerRender, dragActive, defaultDragElement]);

  var handleDrop = function handleDrop(event) {
    var _props$onChange, _props$onDrop;

    event.preventDefault();
    (_props$onChange = props.onChange) === null || _props$onChange === void 0 ? void 0 : _props$onChange.call(props, event.dataTransfer.files);
    (_props$onDrop = props.onDrop) === null || _props$onDrop === void 0 ? void 0 : _props$onDrop.call(props, event);
    setDragActive(false);
  };

  var handleDragenter = function handleDragenter(event) {
    var _props$onDragenter;

    event.preventDefault();
    (_props$onDragenter = props.onDragenter) === null || _props$onDragenter === void 0 ? void 0 : _props$onDragenter.call(props, event);
    setDragActive(true);
  };

  var handleDragleave = function handleDragleave(event) {
    var _props$onDragleave;

    if (!target.current) return;
    event.preventDefault();
    (_props$onDragleave = props.onDragleave) === null || _props$onDragleave === void 0 ? void 0 : _props$onDragleave.call(props, event);
    setDragActive(false);
  };

  var handleDragover = function handleDragover(event) {
    event.preventDefault();
  };

  return /* @__PURE__ */React.createElement("div", {
    ref: target,
    className: classes,
    onDrop: handleDrop,
    onDragEnter: handleDragenter,
    onDragOver: handleDragover,
    onDragLeave: handleDragleave
  }, dragElement);
};

export { Dragger as default };
