// tdesign-react/esm/select-input/useOverlayStyle.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-08260abc.js';
import { _ as _typeof } from '../_chunks/dep-4705ff31.js';
import { _ as _slicedToArray } from '../_chunks/dep-744640f5.js';
import { useState, useMemo } from 'react';
import { i as isObject_1 } from '../_chunks/dep-38424935.js';
import { i as isFunction_1 } from '../_chunks/dep-531703d1.js';
import '../_chunks/dep-0e5be0b2.js';
import '../_chunks/dep-12690e1d.js';
import '../_chunks/dep-21baa91f.js';
import '../_chunks/dep-ad150068.js';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var MAX_POPUP_WIDTH = 1e3;
function useOverlayStyle(props) {
  var popupProps = props.popupProps,
      autoWidth = props.autoWidth,
      readonly = props.readonly,
      onPopupVisibleChange = props.onPopupVisibleChange;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      innerPopupVisible = _useState2[0],
      setInnerPopupVisible = _useState2[1];

  var matchWidthFunc = function matchWidthFunc(triggerElement, popupElement) {
    if (!triggerElement || !popupElement) return;
    var SCROLLBAR_WIDTH = popupElement.scrollHeight > popupElement.offsetHeight ? 8 : 0;
    var width = popupElement.offsetWidth + SCROLLBAR_WIDTH >= triggerElement.offsetWidth ? popupElement.offsetWidth : triggerElement.offsetWidth;
    var otherOverlayStyle = {};

    if (popupProps && _typeof(popupProps.overlayStyle) === "object" && !popupProps.overlayStyle.width) {
      otherOverlayStyle = popupProps.overlayStyle;
    }

    return _objectSpread({
      width: "".concat(Math.min(width, MAX_POPUP_WIDTH), "px")
    }, otherOverlayStyle);
  };

  var onInnerPopupVisibleChange = function onInnerPopupVisibleChange(visible, context) {
    if (readonly) return;
    var newVisible = context.trigger === "trigger-element-click" ? true : visible;
    setInnerPopupVisible(newVisible);
    onPopupVisibleChange === null || onPopupVisibleChange === void 0 ? void 0 : onPopupVisibleChange(newVisible, context);
  };

  var tOverlayStyle = useMemo(function () {
    var result = {};
    var overlayStyle = (popupProps === null || popupProps === void 0 ? void 0 : popupProps.overlayStyle) || {};

    if (isFunction_1(overlayStyle) || isObject_1(overlayStyle) && overlayStyle.width) {
      result = overlayStyle;
    } else if (!autoWidth) {
      result = matchWidthFunc;
    }

    return result;
  }, [autoWidth, popupProps === null || popupProps === void 0 ? void 0 : popupProps.overlayStyle]);
  return {
    tOverlayStyle: tOverlayStyle,
    innerPopupVisible: innerPopupVisible,
    onInnerPopupVisibleChange: onInnerPopupVisibleChange
  };
}

export { useOverlayStyle as default };
