// tdesign-react/es/upload/themes/flow-list/file-list.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../../_chunks/dep-744640f5.js';
import React from 'react';
import { CheckCircleFilledIcon, ErrorCircleFilledIcon, TimeFilledIcon } from 'tdesign-icons-react';
import { abridgeName, returnFileSize } from '../../util.js';
import useConfig from '../../../_util/useConfig.js';
import { useLocaleReceiver } from '../../../locale/LocalReceiver.js';
import { Loading } from '../../../loading/index.js';
import '../../../_chunks/dep-0e5be0b2.js';
import '../../../_chunks/dep-e3ed7551.js';
import '../../../_chunks/dep-08260abc.js';
import '../../../_chunks/dep-12690e1d.js';
import '../../../_chunks/dep-ac0cc4dd.js';
import '../../../_chunks/dep-e3ab6b0d.js';
import '../../../_chunks/dep-cc4f2bae.js';
import '../../../_chunks/dep-531703d1.js';
import '../../../_chunks/dep-21baa91f.js';
import '../../../_chunks/dep-4705ff31.js';
import '../../../_chunks/dep-38424935.js';
import '../../../_chunks/dep-ad150068.js';
import '../../../_chunks/dep-d5f141ac.js';
import '../../../_chunks/dep-1b27e5c2.js';
import '../../../_chunks/dep-6963df41.js';
import '../../../_chunks/dep-8fb1642c.js';
import '../../../_chunks/dep-bcb59ae2.js';
import '../../../_chunks/dep-12718f01.js';
import '../../../_chunks/dep-09c318fb.js';
import '../../../_chunks/dep-048c506f.js';
import '../../../_chunks/dep-febf8cf9.js';
import '../../../_chunks/dep-074143e4.js';
import '../../../_chunks/dep-a1fd625f.js';
import '../../../_chunks/dep-f5ab8a5c.js';
import '../../../_chunks/dep-67009882.js';
import '../../../locale/zh_CN.js';
import '../../../_common/js/global-config/locale/zh_CN.js';
import '../../../config-provider/zh_CN_config.js';
import '../../../config-provider/index.js';
import '../../../config-provider/ConfigProvider.js';
import '../../../loading/Loading.js';
import 'classnames';
import '../../../_util/dom.js';
import 'raf';
import '../../../_chunks/dep-60c49e74.js';
import '../../../_util/easing.js';
import '../../../common/Portal.js';
import 'react-dom';
import '../../../loading/gradient.js';
import '../../../_common/js/loading/circle-adapter.js';
import '../../../_common/js/utils/set-style.js';
import '../../../_common/js/utils/helper.js';
import '../../../_chunks/dep-8ff4a741.js';
import '../../../loading/defaultProps.js';
import '../../../loading/plugin.js';
import './style/css.js';
import '../../../loading/type.js';

var FileList = function FileList(props) {
  var listFiles = props.listFiles,
      showInitial = props.showInitial,
      renderDragger = props.renderDragger,
      showUploadProgress = props.showUploadProgress,
      remove = props.remove;

  var _useConfig = useConfig(),
      prefix = _useConfig.classPrefix;

  var _useLocaleReceiver = useLocaleReceiver("upload"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      locale = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var UPLOAD_NAME = "".concat(prefix, "-upload");
  var progressText = locale.progress,
      infoText = locale.file,
      triggerUploadText = locale.triggerUploadText;

  var renderStatus = function renderStatus(file) {
    var STATUS_MAP = {
      success: {
        icon: /* @__PURE__ */React.createElement(CheckCircleFilledIcon, null),
        text: t(progressText.successText)
      },
      fail: {
        icon: /* @__PURE__ */React.createElement(ErrorCircleFilledIcon, null),
        text: t(progressText.failText)
      },
      progress: {
        icon: /* @__PURE__ */React.createElement(Loading, null),
        text: "".concat(t(progressText.uploadingText), " ").concat(Math.min(file.percent, 99), "%")
      },
      waiting: {
        icon: /* @__PURE__ */React.createElement(TimeFilledIcon, null),
        text: t(progressText.waitingText)
      }
    };
    if (file.status === "progress" && !showUploadProgress) return null;
    return /* @__PURE__ */React.createElement("div", {
      className: "".concat(UPLOAD_NAME, "__flow-status")
    }, STATUS_MAP[file.status].icon, /* @__PURE__ */React.createElement("span", null, STATUS_MAP[file.status].text));
  };

  return /* @__PURE__ */React.createElement("table", {
    className: "".concat(UPLOAD_NAME, "__flow-table")
  }, /* @__PURE__ */React.createElement("thead", null, /* @__PURE__ */React.createElement("tr", null, /* @__PURE__ */React.createElement("th", null, t(infoText.fileNameText)), /* @__PURE__ */React.createElement("th", null, t(infoText.fileSizeText)), /* @__PURE__ */React.createElement("th", null, t(infoText.fileStatusText)), /* @__PURE__ */React.createElement("th", null, t(infoText.fileOperationText)))), /* @__PURE__ */React.createElement("tbody", null, showInitial && /* @__PURE__ */React.createElement("tr", null, /* @__PURE__ */React.createElement("td", {
    colSpan: 4
  }, renderDragger())), listFiles.map(function (file, index) {
    return /* @__PURE__ */React.createElement("tr", {
      key: index
    }, /* @__PURE__ */React.createElement("td", null, abridgeName(file.name, 7, 10)), /* @__PURE__ */React.createElement("td", null, returnFileSize(file.size)), /* @__PURE__ */React.createElement("td", null, renderStatus(file)), /* @__PURE__ */React.createElement("td", null, /* @__PURE__ */React.createElement("span", {
      className: "".concat(UPLOAD_NAME, "__flow-button"),
      onClick: function onClick(e) {
        return remove({
          e: e,
          index: index,
          file: file
        });
      }
    }, t(triggerUploadText["delete"]))));
  })));
};

export { FileList as default };
