// tdesign-react/esm/upload/hooks/useSizeLimit.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-744640f5.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import '../../_chunks/dep-0e5be0b2.js';
import '../../_chunks/dep-08260abc.js';
import 'react';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../_chunks/dep-12690e1d.js';
import '../../_chunks/dep-e3ed7551.js';
import '../../_chunks/dep-ac0cc4dd.js';
import '../../_chunks/dep-e3ab6b0d.js';
import '../../_chunks/dep-cc4f2bae.js';
import '../../_chunks/dep-531703d1.js';
import '../../_chunks/dep-21baa91f.js';
import '../../_chunks/dep-4705ff31.js';
import '../../_chunks/dep-38424935.js';
import '../../_chunks/dep-ad150068.js';
import '../../_chunks/dep-d5f141ac.js';
import '../../_chunks/dep-1b27e5c2.js';
import '../../_chunks/dep-6963df41.js';
import '../../_chunks/dep-8fb1642c.js';
import '../../_chunks/dep-bcb59ae2.js';
import '../../_chunks/dep-12718f01.js';
import '../../_chunks/dep-09c318fb.js';
import '../../_chunks/dep-048c506f.js';
import '../../_chunks/dep-febf8cf9.js';
import '../../_chunks/dep-074143e4.js';
import '../../_chunks/dep-a1fd625f.js';
import '../../_chunks/dep-f5ab8a5c.js';
import '../../_chunks/dep-67009882.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../config-provider/zh_CN_config.js';

var SIZE_MAP = {
  B: 1,
  KB: 1024,
  MB: 1048576,
  GB: 1073741824
};

function judgeIsSuitableSize(fileSize, sizeLimit, unit) {
  var units = ["B", "KB", "MB", "GB"];
  var index = units.indexOf(unit);

  if (index === -1) {
    console.warn("TDesign Upload Warn: `sizeLimit.unit` can only be one of ".concat(units.join()));
  }

  return fileSize <= sizeLimit * SIZE_MAP[unit];
}

function useHandleLimit() {
  var _useLocaleReceiver = useLocaleReceiver("upload"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  return handleSizeLimit;

  function handleSizeLimit() {
    var fileSize = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
    var rawSizeLimit = arguments.length > 1 ? arguments[1] : undefined;
    var sizeLimit = typeof rawSizeLimit === "number" ? {
      size: rawSizeLimit,
      unit: "KB"
    } : rawSizeLimit;
    sizeLimit.unit = sizeLimit.unit.toUpperCase();
    var isSuitableSize = judgeIsSuitableSize(fileSize, sizeLimit.size, sizeLimit.unit);
    var errorMsg;

    if (!isSuitableSize) {
      errorMsg = sizeLimit.message ? t(sizeLimit.message, {
        sizeLimit: sizeLimit.size
      }) : "".concat(t(local.sizeLimitMessage, {
        sizeLimit: sizeLimit.size
      }), " ").concat(sizeLimit.unit);
    }

    return [isSuitableSize, errorMsg];
  }
}

export { useHandleLimit as default };
