// tdesign-react/esm/color-picker/components/panel/header.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import React from 'react';
import classNames from 'classnames';
import { CloseIcon } from 'tdesign-icons-react';
import { COLOR_MODES } from '../../const.js';
import { Radio } from '../../../radio/index.js';
import '../../../_chunks/dep-f265ccc5.js';
import '../../../_chunks/dep-08260abc.js';
import '../../../_util/forwardRefWithStatics.js';
import 'hoist-non-react-statics';
import '../../../common/Check.js';
import '../../../_chunks/dep-744640f5.js';
import '../../../_chunks/dep-0e5be0b2.js';
import '../../../_chunks/dep-8ff4a741.js';
import '../../../_chunks/dep-12690e1d.js';
import '../../../_chunks/dep-21baa91f.js';
import '../../../_chunks/dep-4705ff31.js';
import '../../../_chunks/dep-12718f01.js';
import '../../../_util/helper.js';
import '../../../_chunks/dep-9be29560.js';
import '../../../_chunks/dep-e8f6f682.js';
import '../../../_chunks/dep-a1fd625f.js';
import '../../../_chunks/dep-863396c0.js';
import '../../../_chunks/dep-0db6b692.js';
import '../../../_chunks/dep-6d2697ef.js';
import '../../../_chunks/dep-42701b6c.js';
import '../../../_util/useConfig.js';
import '../../../_chunks/dep-e3ed7551.js';
import '../../../_chunks/dep-ac0cc4dd.js';
import '../../../_chunks/dep-e3ab6b0d.js';
import '../../../_chunks/dep-cc4f2bae.js';
import '../../../_chunks/dep-531703d1.js';
import '../../../_chunks/dep-38424935.js';
import '../../../_chunks/dep-ad150068.js';
import '../../../_chunks/dep-d5f141ac.js';
import '../../../_chunks/dep-1b27e5c2.js';
import '../../../_chunks/dep-6963df41.js';
import '../../../_chunks/dep-8fb1642c.js';
import '../../../_chunks/dep-bcb59ae2.js';
import '../../../_chunks/dep-09c318fb.js';
import '../../../_chunks/dep-048c506f.js';
import '../../../_chunks/dep-febf8cf9.js';
import '../../../_chunks/dep-074143e4.js';
import '../../../_chunks/dep-f5ab8a5c.js';
import '../../../_chunks/dep-67009882.js';
import '../../../locale/zh_CN.js';
import '../../../_common/js/global-config/locale/zh_CN.js';
import '../../../config-provider/zh_CN_config.js';
import '../../../hooks/useControlled.js';
import '../../../_util/noop.js';
import '../../../_util/useCommonClassName.js';
import '../../../radio/defaultProps.js';
import '../../../_util/useMutationObserver.js';
import '../../../_chunks/dep-0a541e21.js';
import '../../../_chunks/dep-5ed520e8.js';
import '../../../radio/style/index.js';
import '../../../radio/type.js';

var Header = function Header(props) {
  var baseClassName = props.baseClassName,
      _props$mode = props.mode,
      mode = _props$mode === void 0 ? "monochrome" : _props$mode,
      colorModes = props.colorModes,
      togglePopup = props.togglePopup,
      _props$closeBtn = props.closeBtn,
      closeBtn = _props$closeBtn === void 0 ? true : _props$closeBtn,
      onModeChange = props.onModeChange;

  var handleClosePopup = function handleClosePopup() {
    togglePopup === null || togglePopup === void 0 ? void 0 : togglePopup(false);
  };

  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(baseClassName, "__head")
  }, /* @__PURE__ */React.createElement("div", {
    className: "".concat(baseClassName, "__mode")
  }, (colorModes === null || colorModes === void 0 ? void 0 : colorModes.length) === 1 ? COLOR_MODES[colorModes[0]] : /* @__PURE__ */React.createElement(Radio.Group, {
    variant: "default-filled",
    size: "small",
    value: mode,
    onChange: onModeChange
  }, Object.keys(COLOR_MODES).map(function (key) {
    return /* @__PURE__ */React.createElement(Radio.Button, {
      key: key,
      value: key
    }, COLOR_MODES[key]);
  }))), closeBtn ? /* @__PURE__ */React.createElement("span", {
    role: "button",
    className: classNames("".concat(baseClassName, "__icon"), "".concat(baseClassName, "__close")),
    onClick: handleClosePopup
  }, /* @__PURE__ */React.createElement(CloseIcon, null)) : null);
};

var PanelHeader = /*#__PURE__*/React.memo(Header);

export { PanelHeader as default };
