// tdesign-react/lib/upload/util.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function returnFileSize(number) {
  if (number < 1024) {
    return "".concat(number, " Bytes");
  }

  if (number >= 1024 && number < 1048576) {
    return "".concat((number / 1024).toFixed(1), " KB");
  }

  if (number >= 1048576) {
    return "".concat((number / 1048576).toFixed(1), " MB");
  }
}
function getCurrentDate() {
  var d = new Date();
  var month = d.getMonth() + 1;
  month = month < 10 ? "0".concat(month) : month;
  return "".concat(d.getFullYear(), "-").concat(month, "-").concat(d.getDate(), " ").concat(d.getHours(), ":").concat(d.getMinutes(), ":").concat(d.getSeconds());
}
function abridgeName() {
  var inputName = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
  var leftCount = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 5;
  var rightcount = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 7;
  var name = inputName;
  var leftLength = 0;
  var rightLength = 0;

  for (var i = 0; i < name.length; i++) {
    var w = name[i];
    var isCn = escape(w).indexOf("%u") === 0;

    if (i < leftCount * 2 && leftLength < leftCount) {
      isCn ? leftLength += 1 : leftLength += 2;
    } else if (i > i - rightcount && rightLength < rightcount) {
      isCn ? rightLength += 1 : rightLength += 2;
    }
  }

  return name.replace(new RegExp("^(.{".concat(leftLength, "})(.+)(.{").concat(rightLength, "})$")), "$1\u2026$3");
}
var urlCreator = function urlCreator() {
  return window.webkitURL || window.URL;
};
function createFileURL(file) {
  var _urlCreator;

  return (_urlCreator = urlCreator()) === null || _urlCreator === void 0 ? void 0 : _urlCreator.createObjectURL(file);
}
function finishUpload(state) {
  return ["success", "fail"].includes(state);
}
function isSingleFile() {
  var multiple = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
  var theme = arguments.length > 1 ? arguments[1] : undefined;
  return !multiple && ["file", "file-input", "image"].includes(theme);
}

exports.abridgeName = abridgeName;
exports.createFileURL = createFileURL;
exports.finishUpload = finishUpload;
exports.getCurrentDate = getCurrentDate;
exports.isSingleFile = isSingleFile;
exports.returnFileSize = returnFileSize;
exports.urlCreator = urlCreator;
