// tdesign-react/lib/table/hooks/useClassName.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _util_useConfig = require('../../_util/useConfig.js');
require('react');
require('../../_chunks/dep-38a46f48.js');
require('../../_chunks/dep-a1655de1.js');
require('../../_chunks/dep-8a149e9f.js');
require('../../_chunks/dep-3b69722d.js');
require('../../_chunks/dep-fde4d28a.js');
require('../../_chunks/dep-e266830c.js');
require('../../_chunks/dep-fcf32c81.js');
require('../../_chunks/dep-3405bda4.js');
require('../../_chunks/dep-15a4424e.js');
require('../../_chunks/dep-c40d75ac.js');
require('../../_chunks/dep-910dcfa3.js');
require('../../_chunks/dep-633a5671.js');
require('../../_chunks/dep-9c339ffc.js');
require('../../_chunks/dep-b1a67107.js');
require('../../_chunks/dep-ab43c961.js');
require('../../_chunks/dep-9cacaf01.js');
require('../../_chunks/dep-c72c7229.js');
require('../../_chunks/dep-8e0a771d.js');
require('../../_chunks/dep-c90cc28f.js');
require('../../_chunks/dep-542cc308.js');
require('../../_chunks/dep-c6637e11.js');
require('../../_common/js/global-config/locale/zh_CN.js');
require('../../config-provider/zh_CN_config.js');

function useClassName() {
  var _useConfig = _util_useConfig["default"](),
      classPrefix = _useConfig.classPrefix;

  var classNames = {
    tableBaseClass: {
      table: "".concat(classPrefix, "-table"),
      body: "".concat(classPrefix, "-table__body"),
      overflowVisible: "".concat(classPrefix, "-table--overflow-visible"),
      content: "".concat(classPrefix, "-table__content"),
      topContent: "".concat(classPrefix, "-table__top-content"),
      bottomContent: "".concat(classPrefix, "-table__bottom-content"),
      tdLastRow: "".concat(classPrefix, "-table__td-last-row"),
      tdFirstCol: "".concat(classPrefix, "-table__td-first-col"),
      thCellInner: "".concat(classPrefix, "-table__th-cell-inner"),
      bordered: "".concat(classPrefix, "-table--bordered"),
      striped: "".concat(classPrefix, "-table--striped"),
      hover: "".concat(classPrefix, "-table--hoverable"),
      loading: "".concat(classPrefix, "-table--loading"),
      rowspanAndColspan: "".concat(classPrefix, "-table--rowspan-colspan"),
      empty: "".concat(classPrefix, "-table__empty"),
      emptyRow: "".concat(classPrefix, "-table__empty-row"),
      headerFixed: "".concat(classPrefix, "-table--header-fixed"),
      columnFixed: "".concat(classPrefix, "-table--column-fixed"),
      widthOverflow: "".concat(classPrefix, "-table--width-overflow"),
      multipleHeader: "".concat(classPrefix, "-table--multiple-header"),
      affixedHeader: "".concat(classPrefix, "-table--affixed-header"),
      affixedHeaderElm: "".concat(classPrefix, "-table__affixed-header-elm"),
      scrollbarDivider: "".concat(classPrefix, "-table__scroll-bar-divider"),
      fullHeight: "".concat(classPrefix, "-table--full-height")
    },
    tdAlignClasses: {
      left: "".concat(classPrefix, "-align-left"),
      right: "".concat(classPrefix, "-align-right"),
      center: "".concat(classPrefix, "-align-center")
    },
    tableHeaderClasses: {
      header: "".concat(classPrefix, "-table__header"),
      thBordered: "".concat(classPrefix, "-table__header-th--bordered"),
      fixed: "".concat(classPrefix, "-table__header--fixed"),
      multipleHeader: "".concat(classPrefix, "-table__header--multiple")
    },
    tableFooterClasses: {
      footer: "".concat(classPrefix, "-table__footer"),
      fixed: "".concat(classPrefix, "-table__footer--fixed")
    },
    tableAlignClasses: {
      top: "".concat(classPrefix, "-vertical-align-top"),
      middle: "".concat(classPrefix, "-vertical-align-middle"),
      bottom: "".concat(classPrefix, "-vertical-align-bottom")
    },
    tableRowFixedClasses: {
      top: "".concat(classPrefix, "-table__row--fixed-top"),
      bottom: "".concat(classPrefix, "-table__row--fixed-bottom"),
      firstBottom: "".concat(classPrefix, "-table__row--fixed-bottom-first"),
      withoutBorderBottom: "".concat(classPrefix, "-table__row--without-border-bottom")
    },
    tableColFixedClasses: {
      left: "".concat(classPrefix, "-table__cell--fixed-left"),
      right: "".concat(classPrefix, "-table__cell--fixed-right"),
      lastLeft: "".concat(classPrefix, "-table__cell--fixed-left-last"),
      firstRight: "".concat(classPrefix, "-table__cell--fixed-right-first"),
      leftShadow: "".concat(classPrefix, "-table__content--scrollable-to-left"),
      rightShadow: "".concat(classPrefix, "-table__content--scrollable-to-right")
    },
    tableLayoutClasses: {
      auto: "".concat(classPrefix, "-table--layout-auto"),
      fixed: "".concat(classPrefix, "-table--layout-fixed")
    },
    tdEllipsisClass: "".concat(classPrefix, "-table-td--ellipsis"),
    tableFullRowClasses: {
      base: "".concat(classPrefix, "-table__row--full"),
      innerFullRow: "".concat(classPrefix, "-table__row-full-inner"),
      innerFullElement: "".concat(classPrefix, "-table__row-full-element"),
      firstFullRow: "".concat(classPrefix, "-table__first-full-row"),
      lastFullRow: "".concat(classPrefix, "-table__last-full-row")
    },
    tableExpandClasses: {
      iconBox: "".concat(classPrefix, "-table__expand-box"),
      iconCell: "".concat(classPrefix, "-table__expandable-icon-cell"),
      row: "".concat(classPrefix, "-table__expanded-row"),
      rowInner: "".concat(classPrefix, "-table__expanded-row-inner"),
      expanded: "".concat(classPrefix, "-table__row--expanded"),
      collapsed: "".concat(classPrefix, "-table__row--collapsed")
    },
    tableSortClasses: {
      sortable: "".concat(classPrefix, "-table__cell--sortable"),
      title: "".concat(classPrefix, "-table__cell--title"),
      trigger: "".concat(classPrefix, "-table__cell--sort-trigger"),
      doubleIcon: "".concat(classPrefix, "-table__double-icons"),
      sortIcon: "".concat(classPrefix, "-table__sort-icon"),
      iconDirection: {
        asc: "".concat(classPrefix, "-table-sort-asc"),
        desc: "".concat(classPrefix, "-table-sort-desc")
      },
      iconActive: "".concat(classPrefix, "-table__sort-icon--active"),
      iconDefault: "".concat(classPrefix, "-icon-sort--default")
    },
    tableSelectedClasses: {
      selected: "".concat(classPrefix, "-table__row--selected"),
      disabled: "".concat(classPrefix, "-table__row--disabled")
    },
    tableFilterClasses: {
      filterable: "".concat(classPrefix, "-table__cell--filterable"),
      popup: "".concat(classPrefix, "-table__filter-pop"),
      icon: "".concat(classPrefix, "-table__filter-icon"),
      popupContent: "".concat(classPrefix, "-table__filter-pop-content"),
      result: "".concat(classPrefix, "-table__filter-result"),
      inner: "".concat(classPrefix, "-table__row-filter-inner"),
      bottomButtons: "".concat(classPrefix, "-table__filter--bottom-buttons"),
      contentInner: "".concat(classPrefix, "-table__filter-pop-content-inner"),
      iconWrap: "".concat(classPrefix, "-table__filter-icon-wrap")
    },
    asyncLoadingClass: "".concat(classPrefix, "-table__async-loading"),
    isFocusClass: "".concat(classPrefix, "-is-focus"),
    isLoadingClass: "".concat(classPrefix, "-is-loading"),
    isLoadMoreClass: "".concat(classPrefix, "-is-load-more"),
    tableTreeClasses: {
      col: "".concat(classPrefix, "-table__tree-col"),
      inlineCol: "".concat(classPrefix, "-table__tree-col--inline"),
      icon: "".concat(classPrefix, "-table__tree-op-icon")
    },
    tableDraggableClasses: {
      rowDraggable: "".concat(classPrefix, "-table--row-draggable"),
      rowHandlerDraggable: "".concat(classPrefix, "-table--row-handler-draggable"),
      colDraggable: "".concat(classPrefix, "-table--col-draggable"),
      handle: "".concat(classPrefix, "-table__handle-draggable"),
      ghost: "".concat(classPrefix, "-table__ele--draggable-ghost"),
      chosen: "".concat(classPrefix, "-table__ele--draggable-chosen"),
      dragging: "".concat(classPrefix, "-table__ele--draggable-dragging")
    },
    virtualScrollClasses: {
      cursor: "".concat(classPrefix, "-table__virtual-scroll-cursor"),
      header: "".concat(classPrefix, "-table__virtual-scroll-header")
    },
    positiveRotate90: "".concat(classPrefix, "-positive-rotate-90"),
    negativeRotate180: "".concat(classPrefix, "-negative-rotate-180")
  };
  return classNames;
}

exports["default"] = useClassName;
