// tdesign-react/lib/color-picker/components/panel/header.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var classNames = require('classnames');
var tdesignIconsReact = require('tdesign-icons-react');
var colorPicker_const = require('../../const.js');
var radio_index = require('../../../radio/index.js');
require('../../../_chunks/dep-9d835441.js');
require('../../../_chunks/dep-a1655de1.js');
require('../../../_util/forwardRefWithStatics.js');
require('hoist-non-react-statics');
require('../../../common/Check.js');
require('../../../_chunks/dep-3cf9becb.js');
require('../../../_chunks/dep-19841e1d.js');
require('../../../_chunks/dep-23535a52.js');
require('../../../_chunks/dep-fcf32c81.js');
require('../../../_chunks/dep-3405bda4.js');
require('../../../_chunks/dep-b1a67107.js');
require('../../../_util/helper.js');
require('../../../_chunks/dep-e3db5727.js');
require('../../../_chunks/dep-86abec96.js');
require('../../../_chunks/dep-c90cc28f.js');
require('../../../_chunks/dep-cb0f33ac.js');
require('../../../_chunks/dep-f3090217.js');
require('../../../_chunks/dep-d4a0b3f0.js');
require('../../../_chunks/dep-6dfb2577.js');
require('../../../_util/useConfig.js');
require('../../../_chunks/dep-38a46f48.js');
require('../../../_chunks/dep-8a149e9f.js');
require('../../../_chunks/dep-3b69722d.js');
require('../../../_chunks/dep-fde4d28a.js');
require('../../../_chunks/dep-e266830c.js');
require('../../../_chunks/dep-15a4424e.js');
require('../../../_chunks/dep-c40d75ac.js');
require('../../../_chunks/dep-910dcfa3.js');
require('../../../_chunks/dep-633a5671.js');
require('../../../_chunks/dep-9c339ffc.js');
require('../../../_chunks/dep-ab43c961.js');
require('../../../_chunks/dep-9cacaf01.js');
require('../../../_chunks/dep-c72c7229.js');
require('../../../_chunks/dep-8e0a771d.js');
require('../../../_chunks/dep-542cc308.js');
require('../../../_chunks/dep-c6637e11.js');
require('../../../_common/js/global-config/locale/zh_CN.js');
require('../../../config-provider/zh_CN_config.js');
require('../../../hooks/useControlled.js');
require('../../../_util/noop.js');
require('../../../_util/useCommonClassName.js');
require('../../../radio/defaultProps.js');
require('../../../_util/useMutationObserver.js');
require('../../../_chunks/dep-85b3eaa4.js');
require('../../../_chunks/dep-e19f22f6.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var classNames__default = /*#__PURE__*/_interopDefaultLegacy(classNames);

var Header = function Header(props) {
  var baseClassName = props.baseClassName,
      _props$mode = props.mode,
      mode = _props$mode === void 0 ? "monochrome" : _props$mode,
      colorModes = props.colorModes,
      togglePopup = props.togglePopup,
      _props$closeBtn = props.closeBtn,
      closeBtn = _props$closeBtn === void 0 ? true : _props$closeBtn,
      onModeChange = props.onModeChange;

  var handleClosePopup = function handleClosePopup() {
    togglePopup === null || togglePopup === void 0 ? void 0 : togglePopup(false);
  };

  return /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassName, "__head")
  }, /* @__PURE__ */React__default["default"].createElement("div", {
    className: "".concat(baseClassName, "__mode")
  }, (colorModes === null || colorModes === void 0 ? void 0 : colorModes.length) === 1 ? colorPicker_const.COLOR_MODES[colorModes[0]] : /* @__PURE__ */React__default["default"].createElement(radio_index.Radio.Group, {
    variant: "default-filled",
    size: "small",
    value: mode,
    onChange: onModeChange
  }, Object.keys(colorPicker_const.COLOR_MODES).map(function (key) {
    return /* @__PURE__ */React__default["default"].createElement(radio_index.Radio.Button, {
      key: key,
      value: key
    }, colorPicker_const.COLOR_MODES[key]);
  }))), closeBtn ? /* @__PURE__ */React__default["default"].createElement("span", {
    role: "button",
    className: classNames__default["default"]("".concat(baseClassName, "__icon"), "".concat(baseClassName, "__close")),
    onClick: handleClosePopup
  }, /* @__PURE__ */React__default["default"].createElement(tdesignIconsReact.CloseIcon, null)) : null);
};

var PanelHeader = /*#__PURE__*/React__default["default"].memo(Header);

exports["default"] = PanelHeader;
