// tdesign-react/lib/date-picker/hooks/useFormat.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dayjs = require('dayjs');
var _common_js_datePicker_utilsNew = require('../../_common/js/date-picker/utils-new.js');
require('../../_chunks/dep-3cf9becb.js');
require('../../_chunks/dep-19841e1d.js');
require('../../_chunks/dep-e9185c09.js');
require('../../_chunks/dep-6dfb2577.js');
require('../../_chunks/dep-c6637e11.js');
require('../../_chunks/dep-3405bda4.js');
require('../../_chunks/dep-c40d75ac.js');
require('../../_chunks/dep-633a5671.js');
require('../../_chunks/dep-e266830c.js');
require('../../_chunks/dep-fcf32c81.js');
require('../../_chunks/dep-15a4424e.js');
require('../../_chunks/dep-9c339ffc.js');
require('../../_chunks/dep-542cc308.js');
require('../../_chunks/dep-d97f1281.js');
require('../../_chunks/dep-e19f22f6.js');
require('../../_chunks/dep-cb0f33ac.js');
require('../../_chunks/dep-b1a67107.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);

var TIME_FORMAT = "HH:mm:ss";
function useFormat(props) {
  var mode = props.mode,
      enableTimePicker = props.enableTimePicker,
      formatFromProps = props.format,
      valueTypeFromProps = props.valueType,
      value = props.value;

  var _getDefaultFormat = getDefaultFormat({
    mode: mode,
    enableTimePicker: enableTimePicker,
    format: formatFromProps,
    valueType: valueTypeFromProps
  }),
      format = _getDefaultFormat.format,
      valueType = _getDefaultFormat.valueType,
      timeFormat = _getDefaultFormat.timeFormat;

  if (enableTimePicker) {
    if (!_common_js_datePicker_utilsNew.extractTimeFormat(format)) console.error("format: ".concat(format, " \u4E0D\u89C4\u8303\uFF0C\u5305\u542B\u65F6\u95F4\u9009\u62E9\u5FC5\u987B\u8981\u6709\u65F6\u95F4\u683C\u5F0F\u5316 HH:mm:ss"));
    if (!_common_js_datePicker_utilsNew.extractTimeFormat(valueType) && valueType !== "time-stamp") console.error("valueType: ".concat(valueType, " \u4E0D\u89C4\u8303\uFF0C\u5305\u542B\u65F6\u95F4\u9009\u62E9\u5FC5\u987B\u8981\u6709\u65F6\u95F4\u683C\u5F0F\u5316 HH:mm:ss"));
  }

  var formatDate = function formatDate(newDate) {
    var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "format";
    var formatMap = {
      format: format,
      valueType: valueType
    };
    var targetFormat = formatMap[type];
    var result;

    if (Array.isArray(newDate)) {
      result = formatRange({
        newDate: newDate,
        format: format,
        targetFormat: targetFormat
      });

      if (result.some(function (r) {
        return r === "Invalid Date";
      })) {
        console.error("\u8BF7\u68C0\u67E5 format\u3001valueType\u3001value \u683C\u5F0F\u662F\u5426\u6709\u6548.\nformat: '".concat(format, "' valueType: '").concat(valueType, "' value: '").concat(value, "'"));
        return [];
      }
    } else {
      result = formatSingle({
        newDate: newDate,
        format: format,
        targetFormat: targetFormat
      });

      if (result === "Invalid Date") {
        console.error("\u8BF7\u68C0\u67E5 format\u3001valueType\u3001value \u683C\u5F0F\u662F\u5426\u6709\u6548.\nformat: '".concat(format, "' valueType: '").concat(valueType, "' value: '").concat(value, "'"));
        return "";
      }
    }

    return result;
  };

  function isValidDate(value2) {
    var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "format";
    var formatMap = {
      format: format,
      valueType: valueType
    };
    var realFormat = formatMap[type];

    if (Array.isArray(value2)) {
      if (realFormat === "time-stamp") return value2.every(function (v) {
        return dayjs__default["default"](v).isValid();
      });
      return value2.every(function (v) {
        return dayjs__default["default"](v, realFormat, true).isValid();
      });
    }

    if (realFormat === "time-stamp") return dayjs__default["default"](value2).isValid();
    return dayjs__default["default"](value2, realFormat, true).isValid();
  }

  function formatTime(value2) {
    var result;

    if (Array.isArray(value2)) {
      result = value2.map(function (v) {
        return dayjs__default["default"](v).format(timeFormat);
      });
    } else {
      result = dayjs__default["default"](value2).format(timeFormat);
    }

    return result;
  }

  return {
    format: format,
    valueType: valueType,
    isValidDate: isValidDate,
    timeFormat: timeFormat,
    formatTime: formatTime,
    formatDate: formatDate,
    getDefaultFormat: getDefaultFormat
  };
}

function formatRange(_ref) {
  var newDate = _ref.newDate,
      format = _ref.format,
      targetFormat = _ref.targetFormat;
  if (!newDate || !Array.isArray(newDate)) return [];
  var dayjsDateList = newDate.map(function (d) {
    return d && (dayjs__default["default"](d).isValid() ? dayjs__default["default"](d) : dayjs__default["default"](d, format));
  });

  if (dayjsDateList[0] && dayjsDateList[1] && dayjsDateList[0].toDate().getTime() > dayjsDateList[1].toDate().getTime()) {
    dayjsDateList.fill(dayjsDateList[1]);
  }

  if (targetFormat === "time-stamp") {
    return dayjsDateList.map(function (da) {
      return da && da.toDate().getTime();
    });
  }

  return dayjsDateList.map(function (da) {
    return da && da.format(targetFormat);
  });
}

function formatSingle(_ref2) {
  var newDate = _ref2.newDate,
      format = _ref2.format,
      targetFormat = _ref2.targetFormat;
  if (!newDate) return "";
  var dayJsDate = dayjs__default["default"](newDate).isValid() ? dayjs__default["default"](newDate) : dayjs__default["default"](newDate, format);
  if (targetFormat === "time-stamp") return dayJsDate.toDate().getTime();
  return dayJsDate.format(targetFormat);
}

function getDefaultFormat(_ref3) {
  var _ref3$mode = _ref3.mode,
      mode = _ref3$mode === void 0 ? "date" : _ref3$mode,
      format = _ref3.format,
      valueType = _ref3.valueType,
      enableTimePicker = _ref3.enableTimePicker;

  if (mode === "year") {
    return {
      format: format || "YYYY",
      valueType: valueType || "YYYY",
      timeFormat: _common_js_datePicker_utilsNew.extractTimeFormat(format || "YYYY") || TIME_FORMAT
    };
  }

  if (mode === "month") {
    return {
      format: format || "YYYY-MM",
      valueType: valueType || "YYYY-MM",
      timeFormat: _common_js_datePicker_utilsNew.extractTimeFormat(format || "YYYY-MM") || TIME_FORMAT
    };
  }

  if (mode === "date") {
    return {
      format: format || "YYYY-MM-DD".concat(enableTimePicker ? " HH:mm:ss" : ""),
      valueType: valueType || "YYYY-MM-DD".concat(enableTimePicker ? " HH:mm:ss" : ""),
      timeFormat: _common_js_datePicker_utilsNew.extractTimeFormat(format || "YYYY-MM-DD".concat(enableTimePicker ? " HH:mm:ss" : "")) || TIME_FORMAT
    };
  }
}

exports.TIME_FORMAT = TIME_FORMAT;
exports["default"] = useFormat;
exports.getDefaultFormat = getDefaultFormat;
