// tdesign-react/esm/date-picker/panel/useTableData.js
/**
 * tdesign v0.34.4
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _slicedToArray } from '../../_chunks/dep-744640f5.js';
import { useLocaleReceiver } from '../../locale/LocalReceiver.js';
import { getWeeks, getMonths, getYears, flagActive } from '../../_common/js/date-picker/utils-new.js';
import '../../_chunks/dep-0e5be0b2.js';
import '../../_chunks/dep-08260abc.js';
import 'react';
import '../../config-provider/index.js';
import '../../config-provider/ConfigProvider.js';
import '../../_chunks/dep-12690e1d.js';
import '../../_chunks/dep-e3ed7551.js';
import '../../_chunks/dep-ac0cc4dd.js';
import '../../_chunks/dep-e3ab6b0d.js';
import '../../_chunks/dep-cc4f2bae.js';
import '../../_chunks/dep-531703d1.js';
import '../../_chunks/dep-21baa91f.js';
import '../../_chunks/dep-4705ff31.js';
import '../../_chunks/dep-38424935.js';
import '../../_chunks/dep-ad150068.js';
import '../../_chunks/dep-d5f141ac.js';
import '../../_chunks/dep-1b27e5c2.js';
import '../../_chunks/dep-6963df41.js';
import '../../_chunks/dep-8fb1642c.js';
import '../../_chunks/dep-bcb59ae2.js';
import '../../_chunks/dep-12718f01.js';
import '../../_chunks/dep-09c318fb.js';
import '../../_chunks/dep-048c506f.js';
import '../../_chunks/dep-febf8cf9.js';
import '../../_chunks/dep-074143e4.js';
import '../../_chunks/dep-a1fd625f.js';
import '../../_chunks/dep-f5ab8a5c.js';
import '../../_chunks/dep-67009882.js';
import '../../locale/zh_CN.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../config-provider/zh_CN_config.js';
import '../../_chunks/dep-3d09d1f3.js';
import '../../_chunks/dep-42701b6c.js';
import '../../_chunks/dep-87214e71.js';
import '../../_chunks/dep-5ed520e8.js';
import '../../_chunks/dep-863396c0.js';
import 'dayjs';

function useTableData(props) {
  var _useLocaleReceiver = useLocaleReceiver("datePicker"),
      _useLocaleReceiver2 = _slicedToArray(_useLocaleReceiver, 2),
      local = _useLocaleReceiver2[0],
      t = _useLocaleReceiver2[1];

  var monthLocal = t(local.months);
  var start = props.start,
      end = props.end,
      hoverStart = props.hoverStart,
      hoverEnd = props.hoverEnd,
      year = props.year,
      month = props.month,
      mode = props.mode,
      firstDayOfWeek = props.firstDayOfWeek,
      disableDate = props.disableDate,
      minDate = props.minDate,
      maxDate = props.maxDate,
      isRange = props.isRange;
  var data = [];
  var options = {
    minDate: minDate,
    maxDate: maxDate,
    disableDate: disableDate,
    firstDayOfWeek: firstDayOfWeek,
    monthLocal: monthLocal
  };

  if (mode === "date") {
    data = getWeeks({
      year: year,
      month: month
    }, options);
  } else if (mode === "month") {
    data = getMonths(year, options);
  } else if (mode === "year") {
    data = getYears(year, options);
  }

  return flagActive(data, {
    start: start,
    end: end,
    hoverStart: hoverStart,
    hoverEnd: hoverEnd,
    type: mode,
    isRange: isRange
  });
}

export { useTableData as default };
